pkg/recommendedversion/recommended_version.go

```diff
@@ -0,0 +1,263 @@
+	alreadyPresent := make(map[string]bool)
+	for _, newVersion := range recommendedArray {
+		trimmedVersion := strings.TrimSpace(newVersion)
+		if trimmedVersion != "" && !alreadyPresent[trimmedVersion] {
+			recommendedVersions = append(recommendedVersions, trimmedVersion)
+			alreadyPresent[trimmedVersion] = true
+		}
+	}
+
+	// Now sort the versions, then reverse the order
+	err := utils.SortVersions(recommendedVersions)
+	if err != nil {
+		return nil, err
+	}
+
+	// Reverse the order so it is descending
+	for i := len(recommendedVersions)/2 - 1; i >= 0; i-- {
+		opp := len(recommendedVersions) - 1 - i
+		recommendedVersions[i], recommendedVersions[opp] = recommendedVersions[opp], recommendedVersions[i]
+	}
+	return recommendedVersions, err
+}
+
+func getRecommendationDelayValue() int {
+	delay := recommendedVersionCheckDelaySeconds
+	delayOverride := os.Getenv(constants.ConfigVariableRecommendVersionDelay)
+	if delayOverride != "" {
+		delayOverrideValue, err := strconv.Atoi(delayOverride)
+		if err == nil && delayOverrideValue >= 0 {
+			delay = delayOverrideValue
+		}
+	}
+	return delay
+}
+
+func shouldCheckVersion() bool {
+	delay := getRecommendationDelayValue()
+	if delay == 0 {
+		// The user has disabled the version check
+		return false
+	}
+
+	// Get the last time the version check was done
+	lastCheck := datastore.GetDataStoreValue(dataStoreLastVersionCheckKey)
+	if lastCheck == nil {
+		return true
+	}
+
+	lastCheckTime, ok := lastCheck.(time.Time)
+	if !ok {
+		return true
+	}
+
+	return time.Since(lastCheckTime) > time.Duration(delay)*time.Second
+}
+
+func printVersionRecommendations(writer io.Writer, currentVersion, major, minor, patch string) {
+	if major == "" && minor == "" && patch == "" {
+		// The current version is the best recommended version
+		return
+	}
+
+	// Put a delimiter before this notification so the user
+	// can see it is not part of the command output
+	fmt.Fprintln(writer, "\n==")
+
+	if utils.IsNewVersion(currentVersion, major) || utils.IsNewVersion(currentVersion, minor) || utils.IsNewVersion(currentVersion, patch) {
+		fmt.Fprintf(writer, "WARNING: Due to a problem it is recommended not to use the current version: %s.\n", currentVersion)
+		fmt.Fprintln(writer, "Please use a recommended version:")
+	} else {
+		fmt.Fprintf(writer, "Note: A new version of the Tanzu CLI is available. You are at version: %s.\n", currentVersion)
+		fmt.Fprintln(writer, "To benefit from the latest security and features, please update to a recommended version:")
+	}
+
+	if major != "" {
+		if utils.IsNewVersion(major, currentVersion) {
+			fmt.Fprintf(writer, "  - %s (new features but breaking changes)\n", major)
+		} else {
+			fmt.Fprintf(writer, "  - %s ([!] you should downgrade to a previous major version)\n", major)
+		}
+	}
+	if minor != "" {
+		if utils.IsNewVersion(minor, currentVersion) {
+			fmt.Fprintf(writer, "  - %s (new features with backwards-compatibility)\n", minor)
```

Comment:
This feels too verbose to me. Anyone knowing the Tanzu CLI version policy should understand this and we can get away with just specifying `(new features)` for minor upgrades. WDYT?
```suggestion
			fmt.Fprintf(writer, "  - %s (new features)\n", minor)
```

Reply:
Yeah, that's good for me.

The UX altogether may be too verbose, so don't hesitate to point out what you don't like, even if you don't have a better suggestion to give.

Reply:
Now that we won't recommend a new major version, there will only be two lines at most.
I think @anujc25 you are right, users will be able to understand the difference, so I'll remove the explanations.